### PR TITLE
Add cgroup memory controller

### PR DESCRIPTION
**Pull request by lebauce**
### What does this PR do?

Proactively call GC when memory used is aboved a specified threshold or when memory pressure events are sent by the kernel

### Motivation

Avoid OOM kills

### CODE REVIEW

**Review by djmitche — APPROVED** [review]

**Review by djmitche — CHANGES_REQUESTED** [review]
I believe this needs an update to [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml)

**Comment by djmitche** [comment]
..and perhaps a release note, since this is introducing an enhancement?

..and a rebase, since the dogstatsd size has increased recently :)

**Review by Gui774ume — APPROVED** [review]

File: cmd/system-probe/utils/memory_monitor_linux.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			}
+
+			memProfile, err := ioutil.TempFile(tmpDir, "memcg-pprof-heap")
+			if err != nil {
+				log.Errorf("Failed to generate memory profile: %s", err)
+				return
+			}
+			defer memProfile.Close()
+
+			if err := pprof.WriteHeapProfile(memProfile); err != nil {
```
Comment (brycekahle) [inline cmd/system-probe/utils/memory_monitor_linux.go:69]: Should we log out the location of the memory profile file?

**Comment by lebauce** [comment]
@djmitche Thanks for the review and sorry for the LONG delay. I addressed your comments and updated the config template. I'm not sure about the release notes though as this is mostly for troubleshooting purposes.

File: pkg/config/config_template.yaml [hunk 2]
```diff
@@ -1386,6 +1386,38 @@ api_key:
...
+    ## Enable cgroup memory controller.
+    #
+    # enabled: false
+    #
+    ## @param thresholds - map of strings - optional
+    ## Thresholds and the respective active actions to trigger when
+    ## memory usage is above the specified threshold.
+    ## Threshold can be either an absolute value - such as 500MB or 2GB -
+    ## or a percentage of the cgroup allocated memory such as 50%.
+    ## The action can be either:
```
Comment (alai97) [inline pkg/config/config_template.yaml:1404]: Either implies one or the other (not three options).

```suggestion
    ## The action can be:
```

File: pkg/config/config_template.yaml [hunk 3]
```diff
@@ -1386,6 +1386,38 @@ api_key:
...
+    #
+    # thresholds:
+    #   500MB: gc
+    #   50%: profile
+
+    ## @param pressure_levels - map of strings - optional
+    ## Pressure levels and the respective active actions to trigger when
+    ## memory usage reaches the specified level.
+    ## The pressure level can be either 'low', 'medium' or 'critical'.
+    ## The actions are the same than for thresholds (see above).
```
Comment (alai97) [inline pkg/config/config_template.yaml:1417]:
```suggestion
    ## The pressure level is 'low', 'medium', or 'critical'.
    ## The actions are the same for thresholds (see above).
```

**Review by alai97 — APPROVED** [review]
Looks good, just two copy nits. Should we maybe update `// Copyright 2016-2020 Datadog, Inc.` to `2016-2022 Datadog, Inc.`?

**Review by djmitche — COMMENTED** [review]
The config template change looks OK for agent-core.

But, I want to be sure: is this something we'd like customers to know about and tweak on their own?  Or is it for internal / support use?  If the latter, then including it in the template may not be a good idea.

**Comment by vboulineau** [comment]
@lebauce Would it make sense to add this feature to our own `cgroups` package: https://github.com/DataDog/datadog-agent/pull/9121?

**Comment by lebauce** [comment]
@vboulineau Yes, it totally makes sense.

@djmitche This is for internal / support but it's hidden behind a boolean `InternalProfiling` false by default

@alai97 Applying your suggestions

**Review by djmitche — APPROVED** [review]
config template changes look good

**Pull request merged**